### ISSUE TITLE
New Rule Idea: Buffers rented from buffer pools should be returned

### ISSUE THREAD

**Issue by sebastien-marichal** [SonarSource/sonar-dotnet#9607]
Buffers allocated using buffer pools such as [ArrayPool](https://learn.microsoft.com/en-us/dotnet/api/system.buffers.arraypool-1?view=net-8.0) should be returned after usage.

### Noncompliant example

```cs
void Method(Stream stream)
{
    var buffer = ArrayPool<byte>.Shared.Rent(512); // Noncompliant buffer is not returned
    var len = stream.Read(buffer, 0, 512);
    // ...
}
```

### Compliant example

```cs
void Method(Stream stream)
{
    var buffer = ArrayPool<byte>.Shared.Rent(512); // Compliant
    var len = stream.Read(buffer, 0, 512);
    // ...
    ArrayPool<byte>.Shared.Return(buffer);
}
```

## Notes

[MemorePool](https://learn.microsoft.com/en-us/dotnet/api/system.buffers.memorypool-1?view=net-8.0) does not have a `Return` and needs more investigation.

**Comment by jilles-sg** [comment]
Left quiet here is that the `ArrayPool<byte>.Shared.Return()` call is not in a `finally` block. This is not wrong but how it should be (assuming that an exception is expected to be rare), since it reduces overhead. Not returning arrays to `ArrayPool<T>.Shared` does not cause leaks, but provably never returning a particular array is of course wrong.

`MemoryPool<T>.Rent()` returns an object that implements `IDisposable` and should be disposed of. It should be assumed that not disposing may cause leaks, so `using` or `finally` should be used. Given that the returned object tends to be a new allocation (albeit a small one), there seems little reason to micro-optimize away a `finally` block.

**Comment by pavel-mikula-sonarsource** [comment]
Internal ticket [NET-1556](https://sonarsource.atlassian.net/browse/NET-1556)

[NET-1556]: https://sonarsource.atlassian.net/browse/NET-1556?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by pavel-mikula-sonarsource** [comment]
Moved
